### PR TITLE
Feature: Azure PDF scan

### LINKED ISSUE #265: Update to v3 API to gain PDF OCR functionality
**Is your enhancement related to a problem? Please describe.**
Following on from the work in #111 where we're using the [OCR API from Computer Vision v2.1](https://westus.dev.cognitive.microsoft.com/docs/services/5cd27ec07268f6c679a3e641/operations/56f91f2e778daf14a499f20d), we should look at updating to the Read API (in either Computer Vision [v3.1](https://westcentralus.dev.cognitive.microsoft.com/docs/services/computer-vision-v3-1-ga/operations/5d986960601faab4bf452005) or [v3.2 that's currently in preview](https://westus.dev.cognitive.microsoft.com/docs/services/computer-vision-v3-2-preview-2/operations/5d986960601faab4bf452005)) to gain access to OCR functionality for PDF files.  [Here's additional details on the Read API](https://docs.microsoft.com/en-us/azure/cognitive-services/computer-vision/concept-recognizing-text), most notably that the free tier would only cover the first two pages of a PDF file.

OCR API currently supports input requirements of:
- Supported image formats: JPEG, PNG, GIF, BMP.
- Image file size must be less than 4MB.
- Image dimensions must be between 50 x 50 and 4200 x 4200 pixels, and the image cannot be larger than 10 megapixels.

Read API currently supports input requirements of:
- Supported image formats: JPEG, PNG, BMP, PDF and TIFF.
- Please do note MPO (Multi Picture Objects) embedded JPEG files are not supported.
- For multi-page PDF and TIFF documents:
   - For the free tier, only the first 2 pages are processed.
   - For the paid tier, up to 2,000 pages are processed.
- Image file size must be less than 50 MB (4 MB for the free tier).
- The image/document page dimensions must be at least 50 x 50 pixels and at most 10000 x 10000 pixels.
- The PDF file dimensions must be at most 17 x 17 inches, corresponding to Legal or A3 paper sizes and smaller.

**Describe the solution you'd like**

**Designs**

**Describe alternatives you've considered**

**Additional context**

### PR DESCRIPTION
**Pull request by dinhtungdu**
### Description of the Change
This PR utilizes the Azure Computer Vision Read API to extract text from multi-page PDF files. It supports both textbase and text-heavy image base pdf files.

Because of the Read API design, this feature uses WP Cron to periodically check and grab the result.

### Verification Process

1. Go to ClassifAI > Image Processing.
2. See the new setting `Enable Scanning PDF`.
3. Enable that feature.
4. Upload a PDF file.
5. Right after the file is uploaded, open its media modal, see `Classifai Read PDF` field with a disabled `In progress!` button.
6. Wait for some minutes for API to process the file, check the modal again, see the description field filled with content of the PDF file.
7. Open the attachment detail page, see a new metabox `Classifai PDF Processing` with rescan checkbox.

### Checklist:

- [x] I have read the [**CONTRIBUTING**](/CONTRIBUTING.md) document.
- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my change.
- [ ] All new and existing tests passed.

### CODE REVIEW

File: includes/Classifai/Providers/Azure/ComputerVision.php [hunk 1]
```diff
@@ -210,7 +235,7 @@ public function setup_attachment_meta_box() {
 	 * @param \WP_Post $post The post object.
 	 */
 	public function attachment_data_meta_box( $post ) {
-		$settings   = get_option( 'classifai_computer_vision' );
+		$settings = get_option( 'classifai_computer_vision' );
```
Comment (dkotter) [inline includes/Classifai/Providers/Azure/ComputerVision.php]: Minor note but looks like extra space got removed here, which means the equal signs don't line up anymore

File: includes/Classifai/Providers/Azure/Read.php [hunk 2]
```diff
@@ -0,0 +1,331 @@
+<?php
+/**
+ * Provides OCR detection with the Computer Vision service.
```
Comment (dkotter) [inline includes/Classifai/Providers/Azure/Read.php]: Not sure this description is accurate for this file

File: includes/Classifai/Providers/Azure/Read.php [hunk 3]
```diff
@@ -0,0 +1,331 @@
...
+
+		if ( empty( $body['error'] ) || empty( $body['error']['code'] ) || empty( $body['error']['message'] ) ) {
+			return $this->log_error( new WP_Error( 'unknown_read_error', esc_html__( 'Unknow Read error.', 'classifai' ) ) );
+		}
+
+		return $this->log_error( new WP_Error( $body['error']['code'], $body['error']['message'] ) );
+	}
+
+	/**
+	 * Use WP Cron to preodically check the status of the read operation.
```
Comment (dkotter) [inline includes/Classifai/Providers/Azure/Read.php]: Minor typo here. Should be `periodically`

**Review by dkotter — COMMENTED** [review]
Haven't fully tested this out yet but code looks good. Left a few minor comments

**Comment by dinhtungdu** [comment]
@dkotter Thanks for the head up, I fixed those typo issues.

**Comment by phpbits** [comment]
@jeffpaul @dinhtungdu Confirming that this feature is working as expected. I followed the steps and the PDF file was scanned successfully. See screenshot below:

Screen Shot 2021-07-14 at 1 04 47 AM

**Review by helen — APPROVED** [review]
Nice, it's working! We probably need to get smarter with the scan button in a future release, because you can do something like request a scan, switch items, come back, and request another scan. It should probably check the status any time that button is loaded up, or load the button with AJAX entirely.

**Pull request merged**